agentcfg/reporter.go

```diff
@@ -0,0 +1,102 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package agentcfg
+
+import (
+	"context"
+	"time"
+
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+type Reporter struct {
+	f        Fetcher
+	p        model.BatchProcessor
+	interval time.Duration
+	logger   *logp.Logger
+	resultc  chan Result
+}
+
+func NewReporter(f Fetcher, batchProcessor model.BatchProcessor, interval time.Duration) Reporter {
+	logger := logp.NewLogger("agentcfg")
+	return Reporter{
+		f:        f,
+		p:        batchProcessor,
+		interval: interval,
+		logger:   logger,
+		resultc:  make(chan Result),
+	}
+}
+
+func (r Reporter) Fetch(ctx context.Context, query Query) (Result, error) {
+	result, err := r.f.Fetch(ctx, query)
+	if err != nil {
+		return Result{}, err
+	}
+	// Only report configs when the query etag == current config etag, or
+	// when the agent indicates it has been applied.
+	if err == nil &&
+		(query.Etag == result.Source.Etag || query.MarkAsAppliedByAgent) {
```

Comment:
```suggestion
	if query.Etag == result.Source.Etag || query.MarkAsAppliedByAgent {
```

(we know err == nil due to check above)